_blog/2024-02-14-writing-async-rust-traits.mdx

```diff
@@ -0,0 +1,318 @@
+
+### Using async_trait
+
+Thankfully, there is a crate that we can use to solve this called `async_trait` which alleviates most of our issues when it comes to async in Rust traits. It provides an attribute macro we can use on the top of any trait to allow `async fn` to be used. We can illustrate this using a trait called `ChatSink` for sending messages:
+
+```rust
+#[async_trait::async_trait]
+trait ChatSink {
+    type Item: Clone;
+
+    async fn send_msg(&mut self, msg: Self::Item) -> Result<(), ChatCommErr>;
+}
+
+#[async_trait::async_trait]
+impl<T> ChatSink for SplitSink<WebSocket, T>
+where
+    Self: SinkExt<T> + Send,
+    T: Send + Clone
+{
+    type Item = T;
+    async fn send_message(&mut self, msg: Self::Item) -> Result<(), ChatCommErr> {
+        self.send(msg).await.map_err(|_| ChatCommErr::new())?;
+        Ok(())
+    }
+}
+```
+
+This trait does the same thing as what we did before with the boxed futures, by expanding the `async fn` to an `fn` that returns a boxed future:
+
+```rust
+impl<T> ChatSink for SplitSink<WebSocket, T> {
+    type Item = T;
+
+    fn send<'async_trait>(
+        &'async_trait self, msg: Self::Item
+    ) -> Pin<Box<dyn core::future::Future<Output = Result<(), ChatCommErr>> + Send + 'async_trait>>{
+         where Self: Sync + 'async_trait, {
+            Box::pin(async move { 
+                // your code block goes here
+            })
+        }
+    }
+}
+```
+
+As you can see, we get rid of most of the complexity by using `async_trait`! A very helpful crate to know. However, because it uses dynamic dispatch it also introduces a little overhead due to needing to use function lookup when run. This means that essentially, this crate is not very helpful if you still want to write heapless programs (or it’s too expensive to put things on the heap). While Rust has managed to make serious headway within async Rust, this is still a little bit of a rough edge.
+
+### Returning “impl Future” in a trait
+
+What about if we don’t know what the future looks like or want to allow the user to decide on the final function signature? For example, if they want to return a box-pinned future? Or just a regular future? We can’t use `async fn` here. In this case, we use `impl Future` as the return type:
+
+```rust
+trait HttpService {
+    async fn fetch(&self, url: Url) -> Body;
+}
+
+// The above would turn into...
+
+trait HttpService {
+    fn fetch(&self, url: Url) -> impl Future<Output = Body>;
+}
+```
+
+A recent change in Rust 1.75 allowed Return Position Impl Trait In Trait (RPITIT - RFCC 1522, 3425).  This is great for us, as it means that we can now use Generic Associated Types (GATs) in a trait! Let’s have a look at the following trait which is just like the previous one, but now has a generic in it:
```

Comment:
GATs have been possible since 1.65, only RPITIT is new in 1.75. In general I would say this section is quite complicated, and I feel like going into GATs here is too much, it could be it's own article even.